src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs

```diff
@@ -127,9 +130,9 @@ public override async Task<IEnumerable<NuGetVersion>> GetAllVersionsAsync(
             {
                 cancellationToken.ThrowIfCancellationRequested();
 
-                var packageInfos = await EnsurePackagesAsync(id, cacheContext, logger, cancellationToken);
+                var packageVersions = await GetAvailablePackageVersionsAsync(id, cacheContext, logger, cancellationToken);
 
-                return packageInfos.Keys;
+                return packageVersions;
```

Comment:
Since `_packageVersionsCache` now contains `HashSet` rather than `SortedDictionary`, this means the public API `GetAllVersionsAsync` may no longer be returning a sorted list of versions.  So, changing this risks breaking behaviour.

However, `FindPackageById` is only used by restore, within NuGet.Client.  PM UI, and anything that cares about unlisted package status, does not use this API.  Restore handles this list not being sorted, and all tests appear to be passing, so I think the biggest risk are apps using our NuGet.Protocol package.